subsys/bluetooth/host/conn.c

```diff
@@ -1318,6 +1319,15 @@ void bt_conn_unref(struct bt_conn *conn)
 
 	__ASSERT(old > 0, "Conn reference counter is 0");
 
+#if defined(CONFIG_BT_CONN)
+	/* Slot has been freed and can be taken. No guarantees are made on requests
+	 * to claim connection object as only the first claim will be served.
+	 */
+	if (atomic_get(&conn->ref) == 0) {
```

Comment:
This will be atomically sound only if we compare to `old`. The adv resume code below is not atomically correct.

It may not really matter if this operation is atomic for users that want to allocate connection objects, since in that case there is no harm in suppressing the event if the connection object is re-taken already, and there is no harm in emitting the event twice, since allocation will be idempotent.  But I think for simplicity, we should be atomically correct.

```suggestion
	if (old == 1) {
```

Feel free to make a local variable like `bool deallocated` if that makes it more readable.

Future work: We like to tie reference ownership to pointers. So we should clear the pointer when we move out the reference. This means `atomic_dec(&conn->ref)` should immediately be followed by `conn = NULL`. If we later need to know some properties of the connection object, then we need to store those in local variables before we decrement the reference counter. Accessing the object after moving out the reference is inherently unsafe.

Reply:
FYI - Not only I addressed the change-request, but also I went ahead and refactored bt_conn_unref() to adopt the approach of not attempting access after `atomic_dec()`.